### PR TITLE
Add release notes for 0.222

### PR DESCRIPTION
**Pull request by mayankgarg1990**
Issue #12945

### CODE REVIEW

File: presto-docs/src/main/sphinx/release/release-0.222.rst [hunk 1]
```diff
@@ -0,0 +1,45 @@
+=============
+Release 0.222
+=============
+
+General Changes
+---------------
+* Fix incorrect results when dividing certain decimal numbers.
+* Fix query failure for certain join queries with "Equi criteria are empty,
+  so INNER join should not have PARTITIONED distribution type".
+* Add support for recoverable grouped execution. This can be enabled with the
```
Comment (mayankgarg1990) [inline presto-docs/src/main/sphinx/release/release-0.222.rst]: @shixuan-fan - can you verify if this point and the one below accurately represent your message.

File: presto-docs/src/main/sphinx/release/release-0.222.rst [hunk 2]
```diff
@@ -0,0 +1,45 @@
+=============
+Release 0.222
+=============
+
+General Changes
+---------------
+* Fix incorrect results when dividing certain decimal numbers.
+* Fix query failure for certain join queries with "Equi criteria are empty,
```
Comment (mayankgarg1990) [inline presto-docs/src/main/sphinx/release/release-0.222.rst]: @rschlussel - this point isn't super clear to me - does it mean fixed the error whose message is this.

File: presto-docs/src/main/sphinx/release/release-0.222.rst [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+Bytecode Changes
+----------------
+* Restore bytecode library deleted in version 0.193.
+
+SPI Changes
+-----------
+* Add experimental interface ``ConnectorPlanOptimizer`` to allow connectors to
+  participate in query plan optimization (e.g., filter pushdown). The interface
+  exposes part of the query plan to the connector to rewrite. The optimized plan
+  will be accepted by the engine in execution. This releases allows connectors
```
Comment (mayankgarg1990) [inline presto-docs/src/main/sphinx/release/release-0.222.rst]: @highker can you explain what does "The optimized plan will be accepted by the engine in execution" mean? Did you mean "The optimized plan will be accepted by the engine during execution" ?

File: presto-docs/src/main/sphinx/release/release-0.222.rst [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+  to access filter and table scan. More operations (union, project, aggregation,
```
Comment (mayankgarg1990) [inline presto-docs/src/main/sphinx/release/release-0.222.rst]: I haven't seen till now (in my small history on the project) having release notes have upcoming features. Should we just focus on the parts shipped in this release?

File: presto-docs/src/main/sphinx/release/release-0.222.rst (hunk 2 above)
Comment (tdcmeehan) [inline presto-docs/src/main/sphinx/release/release-0.222.rst]: I wonder if we could more fully describe this and not have to include the error message here

File: presto-docs/src/main/sphinx/release/release-0.222.rst (hunk 1 above)
Comment (tdcmeehan) [inline presto-docs/src/main/sphinx/release/release-0.222.rst]: Can we link to the issue in case someone wants more details?

File: presto-docs/src/main/sphinx/release/release-0.222.rst [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+* Add support for recoverable grouped execution. This can be enabled with the
+  ``recoverable_grouped_execution`` session property.
+* Add support for grouped execution for ScanFilterProject-only queries. This
+  can be enabled with the ``grouped_execution_for_eligible_table_scans`` session
+  property.
+* Add config property ``max-concurrent-materializations`` and session property
+  ``max_concurrent_materializations`` to limit the number of plan sections that
+  will run concurrently when using materialized exchanges.
+* Add support for computing :func:`approx_distinct` over BingTile values.
+* Add :func:`merge_hll` which performs a scalar merge of an array of HyperLogLogs.
```
Comment (tdcmeehan) [inline presto-docs/src/main/sphinx/release/release-0.222.rst]: This needs to be in the imperative mood.

```suggestion
* Add :func:`merge_hll` to merge an array of HyperLogLogs.
```

File: presto-docs/src/main/sphinx/release/release-0.222.rst [hunk 6]
```diff
@@ -0,0 +1,45 @@
+=============
+Release 0.222
+=============
+
+General Changes
+---------------
+* Fix incorrect results when dividing certain decimal numbers.
```
Comment (tdcmeehan) [inline presto-docs/src/main/sphinx/release/release-0.222.rst]: It would be helpful for an administrator to know what the accuracy problem was.

File: presto-docs/src/main/sphinx/release/release-0.222.rst [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+* Add support for computing :func:`approx_distinct` over BingTile values.
+* Add :func:`merge_hll` which performs a scalar merge of an array of HyperLogLogs.
+* Add bitwise shift operations, :func:`bitwise_arithmetic_shift_right`,
+  :func:`bitwise_logical_shift_right` and :func:`bitwise_shift_left`.
+* Add completed and total lifespans on Web UI.
+
+Hive Connector Changes
+----------------------
+* Fix failures for ``information_schema`` queries when a table has an invalid storage format.
+* Improve query execution time over bucketed table with large buckets.
```
Comment (tdcmeehan) [inline presto-docs/src/main/sphinx/release/release-0.222.rst]: Does this mean large # of buckets or buckets whose files are large?

File: presto-docs/src/main/sphinx/release/release-0.222.rst [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+  can be enabled with the ``grouped_execution_for_eligible_table_scans`` session
+  property.
+* Add config property ``max-concurrent-materializations`` and session property
+  ``max_concurrent_materializations`` to limit the number of plan sections that
+  will run concurrently when using materialized exchanges.
+* Add support for computing :func:`approx_distinct` over BingTile values.
+* Add :func:`merge_hll` which performs a scalar merge of an array of HyperLogLogs.
+* Add bitwise shift operations, :func:`bitwise_arithmetic_shift_right`,
+  :func:`bitwise_logical_shift_right` and :func:`bitwise_shift_left`.
+* Add completed and total lifespans on Web UI.
```
Comment (tdcmeehan) [inline presto-docs/src/main/sphinx/release/release-0.222.rst]:
```suggestion
* Add completed and total lifespans on Presto Coordinator UI.
```

File: presto-docs/src/main/sphinx/release/release-0.222.rst [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+
+Hive Connector Changes
+----------------------
+* Fix failures for ``information_schema`` queries when a table has an invalid storage format.
+* Improve query execution time over bucketed table with large buckets.
+* Add config property ``hive.metastore.glue.catalogid`` to configure the Glue catalog ID.
+
+Bytecode Changes
```
Comment (tdcmeehan) [inline presto-docs/src/main/sphinx/release/release-0.222.rst]: I think we can omit this section entirely as it's a transparent change to everyone except Presto developers.

File: presto-docs/src/main/sphinx/release/release-0.222.rst (hunk 1 above)
Comment (rschlussel) [reply presto-docs/src/main/sphinx/release/release-0.222.rst]: Could we have a clearer explanation of what recoverable grouped execution means.  e.g. "automatic retries for failed buckets when using grouped execution"

File: presto-docs/src/main/sphinx/release/release-0.222.rst [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+=============
+
+General Changes
+---------------
+* Fix incorrect results when dividing certain decimal numbers.
+* Fix query failure for certain join queries with "Equi criteria are empty,
+  so INNER join should not have PARTITIONED distribution type".
+* Add support for recoverable grouped execution. This can be enabled with the
+  ``recoverable_grouped_execution`` session property.
+* Add support for grouped execution for ScanFilterProject-only queries. This
```
Comment (rschlussel) [inline presto-docs/src/main/sphinx/release/release-0.222.rst]: Is there a way to say this without referencing an internal operator name?

File: presto-docs/src/main/sphinx/release/release-0.222.rst [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+SPI Changes
+-----------
+* Add experimental interface ``ConnectorPlanOptimizer`` to allow connectors to
+  participate in query plan optimization (e.g., filter pushdown). The interface
+  exposes part of the query plan to the connector to rewrite. The optimized plan
+  will be accepted by the engine in execution. This releases allows connectors
+  to access filter and table scan. More operations (union, project, aggregation,
+  join, etc) will be available in the coming releases. The usage of
+  ``ConnectorPlanOptimizer`` is similar to ``PlanOptimizer`` in presto-main.
+* Rename ``LogicalRowExpressions::TRUE`` and ``LogicalRowExpressions::FALSE`` to
```
Comment (tdcmeehan) [inline presto-docs/src/main/sphinx/release/release-0.222.rst]: Is this contingent on the above feature being announced above?

File: presto-docs/src/main/sphinx/release/release-0.222.rst [hunk 12]
```diff
@@ -0,0 +1,45 @@
...
+* Improve query execution time over bucketed table with large buckets.
+* Add config property ``hive.metastore.glue.catalogid`` to configure the Glue catalog ID.
+
+Bytecode Changes
+----------------
+* Restore bytecode library deleted in version 0.193.
+
+SPI Changes
+-----------
+* Add experimental interface ``ConnectorPlanOptimizer`` to allow connectors to
```
Comment (tdcmeehan) [inline presto-docs/src/main/sphinx/release/release-0.222.rst]: It would be nice to link to an issue or design doc

File: presto-docs/src/main/sphinx/release/release-0.222.rst (hunk 2 above)
Comment (rschlussel) [reply presto-docs/src/main/sphinx/release/release-0.222.rst]: yes, that's what I meant

File: presto-docs/src/main/sphinx/release/release-0.222.rst (hunk 7 above)
Comment (mayankgarg1990) [reply presto-docs/src/main/sphinx/release/release-0.222.rst]: buckets whose files are large.

File: presto-docs/src/main/sphinx/release/release-0.222.rst (hunk 11 above)
Comment (mayankgarg1990) [reply presto-docs/src/main/sphinx/release/release-0.222.rst]: @highker

File: presto-docs/src/main/sphinx/release/release-0.222.rst (hunk 12 above)
Comment (mayankgarg1990) [reply presto-docs/src/main/sphinx/release/release-0.222.rst]: That is a good point. Even I feel that adding details in the release notes is not the best way to talk about the feature.

Comment (mayankgarg1990) [reply presto-docs/src/main/sphinx/release/release-0.222.rst]: @highker - can you point me to the issue ?

File: presto-docs/src/main/sphinx/release/release-0.222.rst (hunk 2 above)
Comment (mayankgarg1990) [reply presto-docs/src/main/sphinx/release/release-0.222.rst]: As Tim pointed out, can we talk about the real issue instead of the error message?

File: presto-docs/src/main/sphinx/release/release-0.222.rst (hunk 1 above)
Comment (mayankgarg1990) [reply presto-docs/src/main/sphinx/release/release-0.222.rst]: @shixuan-fan can you point me to the issue ?

File: presto-docs/src/main/sphinx/release/release-0.222.rst (hunk 6 above)
Comment (mayankgarg1990) [reply presto-docs/src/main/sphinx/release/release-0.222.rst]: @arhimondr

File: presto-docs/src/main/sphinx/release/release-0.222.rst (hunk 9 above)
Comment (mayankgarg1990) [reply presto-docs/src/main/sphinx/release/release-0.222.rst]: @rschlussel - thoughts?

File: presto-docs/src/main/sphinx/release/release-0.222.rst (hunk 2 above)
Comment (rschlussel) [reply presto-docs/src/main/sphinx/release/release-0.222.rst]: There isn't a good way to describe the kinds of queries that would enounter this error, since it mostly depends on internal implementation details.  But what about: 
```
Fix planning failure for certain join queries caused by invalid distribution type.
```

Comment (mayankgarg1990) [reply presto-docs/src/main/sphinx/release/release-0.222.rst]: That sounds good to me.

File: presto-docs/src/main/sphinx/release/release-0.222.rst (hunk 9 above)
Comment (rschlussel) [reply presto-docs/src/main/sphinx/release/release-0.222.rst]: agreed.  I'm not familiar with this change, but it sounds like it's about implementation. if it's not user facing, it shouldn't be in the release notes.  And if it is user facing, it should be written from the perspective of user impact.

File: presto-docs/src/main/sphinx/release/release-0.222.rst (hunk 6 above)
Comment (arhimondr) [reply presto-docs/src/main/sphinx/release/release-0.222.rst]: The result was totally incorrect for some rare cases in division. The overflow of an integer was happening, and instead of an unsigned division - signed division was used.

File: presto-docs/src/main/sphinx/release/release-0.222.rst (hunk 1 above)
Comment (shixuan-fan) [reply presto-docs/src/main/sphinx/release/release-0.222.rst]: The one for `grouped_execution_for_eligible_table_scans` looks good. 

For recoverable grouped execution one, this could work but I'm not sure if we also need to specify the limitation. Maybe we could add `Currently this is only available  for limited query shapes when input and output connectors are Hive connector`. This is still not 100% accurate/clear. The limitation is specified in comment in `PlanFragmenter` and has multiple bullet points, but I think it is not suitable for broad audience of release note. (https://github.com/prestodb/presto/blob/1e3f420e67b5de8fae9b9f407e7f80d94ed508d5/presto-main/src/main/java/com/facebook/presto/sql/planner/PlanFragmenter.java#L216-L225).

I also want to see what the committers think about how we should clarify this without exposing too many unnecessary details.

Comment (shixuan-fan) [reply presto-docs/src/main/sphinx/release/release-0.222.rst]: Related issue: https://github.com/prestodb/presto/issues/12124

File: presto-docs/src/main/sphinx/release/release-0.222.rst (hunk 10 above)
Comment (shixuan-fan) [reply presto-docs/src/main/sphinx/release/release-0.222.rst]: How about `mapper only`? I originally thought about using mapper but then think we don't have such concept. Basically it means "no join and no aggregation".

File: presto-docs/src/main/sphinx/release/release-0.222.rst (hunk 12 above)
Comment (highker) [reply presto-docs/src/main/sphinx/release/release-0.222.rst]: https://github.com/prestodb/presto/pull/12920/commits/fd6ae44d37984dd37d0c3f8055ec435cbb01975b#diff-96c6cc2dde994497e64c36f816baf0c2R23 is the javadoc. That is the only thing we have for now.

File: presto-docs/src/main/sphinx/release/release-0.222.rst (hunk 11 above)
Comment (highker) [reply presto-docs/src/main/sphinx/release/release-0.222.rst]: These two flags were introduced 2 releases earlier. So I think it worth mentioning in case some users have already used them in the connectors.

File: presto-docs/src/main/sphinx/release/release-0.222.rst (hunk 3 above)
Comment (highker) [reply presto-docs/src/main/sphinx/release/release-0.222.rst]: meaning "The optimized plan will be accepted by the engine during planning"

File: presto-docs/src/main/sphinx/release/release-0.222.rst [hunk 13]
```diff
@@ -0,0 +1,42 @@
...
+  can be enabled with the ``grouped_execution_for_eligible_table_scans`` session
+  property.
+* Add config property ``max-concurrent-materializations`` and session property
+  ``max_concurrent_materializations`` to limit the number of plan sections that
+  will run concurrently when using materialized exchanges.
+* Add support for computing :func:`approx_distinct` over BingTile values.
+* Add :func:`merge_hll` to merge of an array of HyperLogLogs.
+* Add bitwise shift operations, :func:`bitwise_arithmetic_shift_right`,
+  :func:`bitwise_logical_shift_right` and :func:`bitwise_shift_left`.
+* Add completed and total lifespans on Presto Coordinator UI.
```
Comment (rongrong) [inline presto-docs/src/main/sphinx/release/release-0.222.rst]: We used to put UI changes into a separate section `Web UI Changes`

File: presto-docs/src/main/sphinx/release/release-0.222.rst (hunk 3 above)
Comment (rongrong) [reply presto-docs/src/main/sphinx/release/release-0.222.rst]: "This releases" -> "This release"

File: presto-docs/src/main/sphinx/release/release-0.222.rst (hunk 10 above)
Comment (rschlussel) [reply presto-docs/src/main/sphinx/release/release-0.222.rst]: What about `single stage queries` (to use the presto concept)?

Comment (mayankgarg1990) [reply presto-docs/src/main/sphinx/release/release-0.222.rst]: But for something like `INSERT INTO <> AS SELECT * FROM foo` wouldn't we have 3 stages? 1) Scan 2) Write 3) Coordinator stats

File: presto-docs/src/main/sphinx/release/release-0.222.rst (hunk 12 above)
Comment (highker) [reply presto-docs/src/main/sphinx/release/release-0.222.rst]: I agree this a big change; documentation is desired. Actually, if there is a way to link to the design doc in slack (https://prestodb.slack.com/archives/C07JH9WMQ/p1555713637042300), that would be good too.

File: presto-docs/src/main/sphinx/release/release-0.222.rst (hunk 10 above)
Comment (rschlussel) [reply presto-docs/src/main/sphinx/release/release-0.222.rst]: I guess that's true. what about `Add support for grouped execution for queries with no joins or aggregations`?

File: presto-docs/src/main/sphinx/release/release-0.222.rst (hunk 6 above)
Comment (mayankgarg1990) [reply presto-docs/src/main/sphinx/release/release-0.222.rst]: Referred to the PR

File: presto-docs/src/main/sphinx/release/release-0.222.rst (hunk 10 above)
Comment (mayankgarg1990) [reply presto-docs/src/main/sphinx/release/release-0.222.rst]: Changed as per Rebecca's suggestion. Also added link to PR.

File: presto-docs/src/main/sphinx/release/release-0.222.rst (hunk 12 above)
Comment (rschlussel) [reply presto-docs/src/main/sphinx/release/release-0.222.rst]: why don't you copy the design doc into a github issue?

Comment (highker) [reply presto-docs/src/main/sphinx/release/release-0.222.rst]: Quickly created on (#13102). Hope that could help

File: presto-docs/src/main/sphinx/release/release-0.222.rst [hunk 14]
```diff
@@ -0,0 +1,44 @@
...
+Release 0.222
+=============
+
+General Changes
+---------------
+
+* Fix incorrect results when dividing certain decimal numbers (:pr:`12930`).
+* Fix planning failure for certain join queries caused by invalid distribution
+  type (:pr:`12354`).
+* Add support for automatic retries for failed buckets when using grouped
```
Comment (rschlussel) [inline presto-docs/src/main/sphinx/release/release-0.222.rst]: A few comments
1. The first sentence is a bit hard to parse
2. Is there anything more specific to describe what's supported other than "certain queries"?
3. We usually start with the configuration properties and then note the functionality https://github.com/prestodb/presto/wiki/Release-Notes-Guidelines#phrasing (though the description there seems like its saying the opposite of what the example does, the example is what I've seen us do in other release notes).  I don't personally care either way, but we should be consistent.

Something like this 
```
Add  session property ``recoverable_grouped_execution`` [and configuration property...] to enable automatically retrying failed buckets when using grouped execution.  Currently this is only supported in the Hive connector and for queries that <whatever characterizes those "certain queries"> (:issue:`12124`).
 ```

File: presto-docs/src/main/sphinx/release/release-0.222.rst [hunk 15]
```diff
@@ -0,0 +1,44 @@
...
+  execution for certain queries when using the Hive connector. This can be
+  enabled with the ``recoverable_grouped_execution`` session property (:issue:`12124`).
+* Add support for grouped execution for queries with no joins or aggregations.
+  This can be enabled with the ``grouped_execution_for_eligible_table_scans``
+  session property (:pr:`12934`).
+* Add config property ``max-concurrent-materializations`` and session property
```
Comment (rschlussel) [inline presto-docs/src/main/sphinx/release/release-0.222.rst]: config -> configuration

File: presto-docs/src/main/sphinx/release/release-0.222.rst [hunk 16]
```diff
@@ -0,0 +1,44 @@
...
+General Changes
+---------------
+
+* Fix incorrect results when dividing certain decimal numbers (:pr:`12930`).
+* Fix planning failure for certain join queries caused by invalid distribution
+  type (:pr:`12354`).
+* Add support for automatic retries for failed buckets when using grouped
+  execution for certain queries when using the Hive connector. This can be
+  enabled with the ``recoverable_grouped_execution`` session property (:issue:`12124`).
+* Add support for grouped execution for queries with no joins or aggregations.
```
Comment (rschlussel) [inline presto-docs/src/main/sphinx/release/release-0.222.rst]: Same thing about putting properties first and also about whether there is a corresponding configuration property. 

```
Add session property ``grouped_execution_for_eligible_table_scans`` [and configuration property...] to enable grouped execution for queries with no joins or aggregations (:pr:`12934`).
```

File: presto-docs/src/main/sphinx/release/release-0.222.rst (hunk 14 above)
Comment (rschlussel) [reply presto-docs/src/main/sphinx/release/release-0.222.rst]: we concluded that if the note is about a new feature and not mostly about configuring the cluster, then the description could go first.  We'll update the release note guidelines to be clearer.
